keyboards/lily58/keymaps/via/keymap.c

```diff
@@ -0,0 +1,163 @@
+ * |------+------+------+------+------+------|                    |------+------+------+------+------+------|
+ * |      |      |      |      |      |      |-------.    ,-------|      |      |RGB ON| HUE+ | SAT+ | VAL+ |
+ * |------+------+------+------+------+------|       |    |       |------+------+------+------+------+------|
+ * |      |      |      |      |      |      |-------|    |-------|      |      | MODE | HUE- | SAT- | VAL- |
+ * `-----------------------------------------/       /     \      \-----------------------------------------'
+ *                   | LAlt | LGUI |LOWER | /Space  /       \Enter \  |RAISE |BackSP| RGUI |
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+  [_ADJUST] = LAYOUT( \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,\
+                             _______, _______, _______, _______, _______,  _______, _______, _______ \
+  )
+};
+
+// Setting ADJUST layer RGB back to default
+void update_tri_layer_RGB(uint8_t layer1, uint8_t layer2, uint8_t layer3) {
+  if (IS_LAYER_ON(layer1) && IS_LAYER_ON(layer2)) {
+    layer_on(layer3);
+  } else {
+    layer_off(layer3);
+  }
+}
+
+//SSD1306 OLED update loop, make sure to enable OLED_DRIVER_ENABLE=yes in rules.mk
+#ifdef OLED_DRIVER_ENABLE
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+  if (!is_keyboard_master())
+    return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
+  return rotation;
+}
+
+// When you add source files to SRC in rules.mk, you can use functions.
+const char *read_layer_state(void);
+const char *read_logo(void);
+void set_keylog(uint16_t keycode, keyrecord_t *record);
+const char *read_keylog(void);
+const char *read_keylogs(void);
+
+// const char *read_mode_icon(bool swap);
+// const char *read_host_led_state(void);
+// void set_timelog(void);
+// const char *read_timelog(void);
+
+void oled_task_user(void) {
+  if (is_keyboard_master()) {
+    // If you want to change the display of OLED, you need to change here
+    oled_write_ln(read_layer_state(), false);
+    oled_write_ln(read_keylog(), false);
+    oled_write_ln(read_keylogs(), false);
+    //oled_write_ln(read_mode_icon(keymap_config.swap_lalt_lgui), false);
+    //oled_write_ln(read_host_led_state(), false);
+    //oled_write_ln(read_timelog(), false);
+  } else {
+    oled_write(read_logo(), false);
+  }
```

Comment:
Honestly, I would much rather that this be converted to use `oled_write_ln_P`, which means that you'd have to remve the SRC for the /lib/ files in the rules.mk, and move the functions into this file. 

This is more work, but is important, IMO, because the `*_P` functions use a lot less memory, which is a factor on split boards with OLED (source, me, as I've run into this issue personally, without adding VIA into the mix).

Reply:
i have not yet addressed this, although I took a stab at it. In my efforts to do so, i was able to reduce firmware size from about 80% to 76%, but i was unable to get the keylogger to work correctly. Layer and logo went over fine.

Reply:
The keylogger won't work with the `*_P` stuff.  The `*_P` is to denote that the strings are in PROGMEM (program memory, aka firmware) rather than in memory.  However, the keylogger MUST be in memory, since the data changes, as PROGMEM cannot be changed (at least, not without flashing). 

So I don't think the "const" is strictly needed, but since the data can't/won't change, it's better to use const. 

Basically, anythign that needs to dynamically change (such as the keylogger stuff) can't be progmem, but everything else can and should be.

Reply:
Also, if you want a good example of what you can do with the OLED stuff, I have a good example in the corne keymap folder: 
https://github.com/qmk/qmk_firmware/blob/master/keyboards/crkbd/keymaps/oled_sample/keymap.c

Reply:
in that case would you prefer that I leave the keylogger as not *_P? or simply remove it?

Reply:
The keylogger must not use the `_P` version.  That is fine, here.

Reply:
okay, leaving the keylogger using _P functions. Currently compiles to use about 80% of available space